### PR TITLE
perf(cast): improve vanity wallet generation

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

`iter::repeat_with`'s function is not executed in parallel and is 20x slower than when called with any other constant time iterator (ranges etc) paired with a parallel map

## Solution

- fix: hex_address_validator in the wrong spot
- perf: use `iter::repeat` (interchangeable with `(0..)`) with a parallel map instead of `iter::repeat_with`

### System info

- Using CPU
  - OS: Ubuntu 22.04.1 LTS x86_64
  - Kernel: 5.15.0-47-generic
  - CPU: AMD Ryzen 9 5900X (24) @ 3.700GHz

- Using GPU
  - OS: Windows 10 Pro
  - Build: 19041.1.amd64fre.vb_release.191206-1406
  - GPU: NVIDIA GeForce RTX 3060

- Rust: latest nightly
- Browser: Brave

---

### [profanity](https://github.com/johguse/profanity) (old README.md benches)
- GTX 1070 OC: 179M/s
- GTX 1070: 163M/s
- RX 480: 120M/s

### [vanity-eth.tk](https://github.com/bokub/vanity-eth) (prefix, case insensitive)
~2K/s per thread, with diminishing returns

### cast wallet vanity
- Local benches:
  - (master) iter::repeat_with, 10K
    - local: in ~1310ms ~= 7.7K/s
  - range iter / iter::repeat + par map, 100K
    - debug: in ~4300ms ~= 23.2K/s
    - local: in ~610ms ~= 165K/s
    - release: in ~440ms ~= 227K/s

- Criterion, using `black_box` so equivalent to debug mode:
  - range iter / iter::repeat + par map
 
 image

---

### [create2crunch](https://github.com/0age/create2crunch) (`-r`, `WORK_SIZE = 0x40000000` (one extra '0'))
CREATE2 salt calculations are way cheaper, but wanted to include it anyway
- CPU: 10M in ~295ms ~= 10M/s
- GPU: ~650M/s

---

### Benchmarking code

```rust
use std::{
    sync::atomic::{AtomicU64, Ordering},
    time::{Duration, SystemTime, UNIX_EPOCH},
};

const STEP: u64 = 100_000;
static PREVIOUS: AtomicU64 = AtomicU64::new(0);
static I: AtomicU64 = AtomicU64::new(0);

#[inline]
fn setup() {
    let start = now().as_millis() as u64;
    PREVIOUS.store(start, Ordering::Release);
}

#[inline]
fn bench() {
    let i = I.fetch_add(1, Ordering::AcqRel);
    if i % STEP == 0 {
        let now = now().as_millis() as u64;
        let previous =
            PREVIOUS.fetch_update(Ordering::Release, Ordering::Acquire, |_| Some(now)).unwrap();
        println!("{} in {} ms", STEP, now - previous);
    };
}

#[inline]
fn now() -> Duration {
    SystemTime::now().duration_since(UNIX_EPOCH).unwrap()
}
```

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
great!

thanks for looking into this and adding benches.

lgtm, only a few nits

File: cli/src/cmd/cast/wallet/vanity.rs [hunk 1]
```diff
@@ -130,41 +128,54 @@ impl Cmd for VanityArgs {
...
-    .find_any(|(_, addr)| matcher.is_match(addr))
-    .map(|(key, _)| key.into())
+type _Wallet = (SigningKey, H160);
+
+pub fn find_vanity_address<T: VanityMatcher>(m: T) -> Option<LocalWallet> {
+    wallet_generator().find_any(matcher(m)).map(|(key, _)| key.into())
 }
 
-fn find_vanity_address_with_nonce<T: VanityMatcher>(matcher: T, nonce: u64) -> Option<LocalWallet> {
+pub fn find_vanity_address_with_nonce<T: VanityMatcher>(m: T, nonce: u64) -> Option<LocalWallet> {
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply cli/src/cmd/cast/wallet/vanity.rs:137]: was due to name collision with the function

**Review by mattsse — APPROVED** [review]
tysm!

**Comment by gakonst** [comment]
sick

**Pull request merged**